tool/lib/license_utils.dart

```diff
@@ -0,0 +1,378 @@
+    required this.includePaths,
+    required this.excludePaths,
+    required this.fileTypes,
+  });
+
+  /// Reads the contents of the yaml [file] and parses it into a [LicenseConfig]
+  /// object.
+  factory LicenseConfig.fromYamlFile(File file) {
+    final yamlString = file.readAsStringSync();
+    final yamlDoc = loadYamlDocument(yamlString);
+    final yaml = yamlDoc.contents as YamlMap;
+    final updatePaths = yaml['update_paths'];
+
+    return LicenseConfig(
+      removeLicenses: yaml['remove_licenses'],
+      addLicenses: yaml['add_licenses'],
+      includePaths: updatePaths['include'],
+      excludePaths: updatePaths['exclude'],
+      fileTypes: updatePaths['file_types'],
+    );
+  }
+
+  /// Sequence of license text strings that should be matched against at the top
+  /// of a file and removed.
+  final YamlList removeLicenses;
+
+  /// Sequence of license text strings that should be added to the top of a
+  /// file.
+  final YamlList addLicenses;
+
+  /// Path(s) to recursively check for file to remove/add license text
+  final YamlList includePaths;
+
+  /// Path(s) to recursively check for files to ignore
+  final YamlList excludePaths;
+
+  /// Contains the extension (without a '.') and the associated indices
+  /// of [removeLicenses] to remove and index of [addLicenses] to add for the
+  /// file type.
+  final YamlMap fileTypes;
+
+  /// Returns the list of indices for the given [ext] of [removeLicenses]
+  /// containing the license text to remove.
+  YamlList getRemoveIndicesForExtension(String ext) {
+    final fileType = fileTypes[_removeDotFromExtension(ext)];
+    return fileType['remove'] as YamlList;
+  }
+
+  /// Returns the index for the given [ext] of [addLicenses] containing the
+  /// license text to add.
+  int getAddIndexForExtension(String ext) {
+    final fileType = fileTypes[_removeDotFromExtension(ext)];
+    return fileType['add'];
+  }
+
+  /// Returns whether the file should be excluded according to the config.
+  bool shouldExclude(File file) {
+    var included = false;
+    for (final includePath in includePaths) {
+      if (file.path.startsWith(includePath)) {
+        included = true;
+        break;
+      }
+    }
+    var excluded = false;
+    for (final excludePath in excludePaths) {
+      if (file.path.startsWith(excludePath)) {
+        excluded = true;
+        break;
+      }
+    }
+    return !included || excluded;
+  }
+
+  /// The config expects an extension without a starting dot, so remove
+  /// any that might exist (e.g. in the extension that is returned by
+  /// [p.extension]).
+  String _removeDotFromExtension(String ext) {
+    if (ext.startsWith('.') && ext.length > 1) {
+      return ext.substring(1);
+    }
+    return ext;
+  }
+}
+
+/// Map containing:
+/// - existing license text for key [LicenseHeader.existingHeaderKey]
+/// - replacement license text for key [LicenseHeader.replacementHeaderKey]
+typedef ReplacementInfo = Map<String, String>;
```

Comment:
On second thought, maybe these should be record types since there's a fixed set of keys?